File: ios/RNBrimodSDK/RNBrimodSDK.m
//
//  RNBrimodSDK.m
//  RNBrimodSDK
//
//  Created by Bangkit's MacBook on 09/07/25.
//

#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_MODULE(RNBrimodSDK, RCTEventEmitter)

RCT_EXTERN_METHOD(selectTabBarItem:
                  (NSNumber * _Nonnull)index
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(requestApiCall:
                  (NSString *)apiName
                  payload:(NSDictionary * _Nullable)payload
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(navigateToNative:
                  (NSString * _Nullable)name
                  params:(NSDictionary * _Nullable)params
                  isRnDismissed:(NSNumber * _Nonnull)isRnDismissed)

RCT_EXTERN_METHOD(navigateToReact:
                  (NSString * _Nullable)bundleName
                  appName(NSString * _Nonnull)appName
                  params:(NSDictionary * _Nullable)params)

RCT_EXTERN_METHOD(dismissReact)

RCT_EXTERN_METHOD(sendDataToReact:
                  (NSString *)data)

RCT_EXTERN_METHOD(sendDataToNative:
                  (NSString *)name
                  data:(NSString *)data)

@end

File: ios/RNBrimodSDK/RNBrimodSDK.swift
//
//  RNBrimodSDK.swift
//  RNBrimodSDK
//
//  Created by Bangkit's MacBook on 08/07/25.
//

import Foundation

public protocol RNBrimodSDKProtocol {
  func setDelegate(_ delegate: RNBrimodSDKProtocol)
  
  func selectTabBarItem(
      _ index: NSNumber,
      onSuccess: @escaping (String) -> Void,
      onError: @escaping (String, String, Error) -> Void
  )
  
  func requestApiCall(
      _ apiName: String,
      payload: [String: Any]?,
      onSuccess: @escaping (String) -> Void,
      onError: @escaping (String, String, Error) -> Void
  )
  
  func navigateToNative(
      _ name: String,
      params: [String: Any]?,
      isRnDismissed: NSNumber
  )
  
  func sendDataToReact(_ data: [String: Any])

  func sendDataToNative(_ name: String, data: [String: Any])
}

@objc(RNBrimodSDK)
class RNBrimodSDK: RCTEventEmitter {
  
  private var delegate: RNBrimodSDKProtocol?
  
  @objc
  func selectTabBarItem(
      _ index: NSNumber,
      resolver: @escaping RCTPromiseResolveBlock,
      rejecter: @escaping RCTPromiseRejectBlock
  ) {
    delegate?.selectTabBarItem(
      index,
      onSuccess: { result in
        resolver(result)
      },
      onError: { code, message, error in
        rejecter(code, message, error)
      }
    )
  }
  
  @objc
  func requestApiCall(
      _ apiName: String,
      payload: [String: Any]?,
      resolver: @escaping RCTPromiseResolveBlock,
      rejecter: @escaping RCTPromiseRejectBlock
  ) {
    delegate?.requestApiCall(
      apiName,
      payload: payload,
      onSuccess: { result in
        resolver(result)
      },
      onError: { code, message, error in
        rejecter(code, message, error)
      }
    )
  }
  
  @objc
  func navigateToNative(
      _ name: String,
      params: [String: Any]?,
      isRnDismissed: NSNumber
  ) {
    delegate?.navigateToNative(name, params: params, isRnDismissed: isRnDismissed)
  }
  
  @objc
  func sendDataToReact(_ data: [String: Any]) {
    delegate?.sendDataToReact(data)
  }

  @objc
  func sendDataToNative(_ name: String, data: [String: Any]) {
    delegate?.sendDataToNative(name, data: data)
  }
}
